src/fparser/block_statements.py

```diff
@@ -821,28 +824,65 @@ def process_item(self):
 
 # SelectCase
 
+
 class EndSelect(EndStatement):
     match = re.compile(r'end\s*select\s*\w*\Z', re.I).match
     blocktype = 'select'
 
+
 class Select(BeginStatement):
     """
-    [ <case-construct-name> : ] SELECT CASE ( <case-expr> )
+    Base class for the Select (case/type) statement
 
     """
-    match = re.compile(r'select\s*case\s*\(.*\)\Z',re.I).match
     end_stmt_cls = EndSelect
     name = ''
-    def tostr(self):
-        return 'SELECT CASE ( %s )' % (self.expr)
+
     def process_item(self):
-        self.expr = self.item.get_line()[6:].lstrip()[4:].lstrip()[1:-1].strip()
+        ''' Populate the state of this Select object by parsing the
+        associated line of code '''
+        item = self.item
+        # TODO make the following more robust, particularly to the
+        # presence of a name at the beginning
+        # (e.g. "a_name: select case(...)")
+        line = item.get_line()[6:].lstrip()[4:].lstrip()[1:-1].strip()
+        self.expr = item.apply_map(line)
```

Comment:
It maps from a Fortran sub-expression to a tuple object. Without this the generated Fortran was wrong if the case statement contained a sub-expression.